Switch python server to use Flask

dbunker opened the pull request.

zpao commented:
Is Flask super common? My original goal here was that the servers would not require any dependencies be installed. I didn't want anybody to really have to think about the server part, it should "just work" so the reader can focus on React and not have to worry. I didn't get there with the node server (mostly because of lack of time and the fact that npm is pretty much required), though I've considered going back to make sure it does.

The server we have works and there's no obvious advantage to using Flask besides that the code is (much) cleaner. But I wouldnt mind being convinced.

dbunker commented:
It's sort of arbitrary what gets included in the standard library of a programming language. C, for example, doesn't include any http server, so unless you want to implement HTTP starting from scratch you're not going to be getting very far. I doubt many people are writing server side code using SimpleHTTPServer, an awful lot are writing code using Flask though, it's quite [popular](https://github.com/mitsuhiko/flask), at least by stargazer count. 

I'd say the main purpose of server side code in a tutorial for client side code is so the reader can immediately understand what the server is doing and then immediately return to the tutorial at hand. I can't say I immediately understood what the SimpleHTTPServer code was doing. Since we're assuming the user will go to whichever they know the best, ruby, node, or python, it's reasonable to assume if they go to python they know what pip is, or at least could google it sufficiently to install Flask. Those who know Flask would quickly know what the code does, for the rest, the code seems clearer, at least to me, than using SimpleHTTPServer.

Also updated the code to reflect that server code writes to _comments.txt. Added pretty printing, though could remove that to match up to the other servers.

spicyj commented:
Flask is probably about as popular as Sinatra. (And as someone who writes^Wused to write Python on a daily basis, Flask is much more familar to me than SimpleHTTPRequestHandler.)

Zopieux commented on server.py:
```diff
@@ -8,57 +8,24 @@
-    print "Server started: http://localhost:3000/"
-    httpd = HTTPServer(('127.0.0.1', 3000), MyHandler)
-    httpd.serve_forever()
+app.run(port=3000)
```
Why remove the `__name__` check? Putting code in the root context like that is a bad practise.

spicyj replied:
(+1)

zpao commented:
Ok, I'm game then. Does pip have anything to describe dependencies, like `package.json` for npm and `Gemfile` for ruby?

zpao commented:
See also, #20, where we're going to make it so the json is reread before every response. We should make sure that functionality doesn't get lost in the switch.

Also, before we can merge this, we'll need the CLA signed (https://code.facebook.com/cla). Our bot wasn't set up right and missed PRs in this repo.

Also also, @spicyj should we switch to using Sinatra for the ruby version? It's one more place where something could go wrong. (it might actually be better to keep 2 versions, including for this... `server_simple.py ` and `server_flask.py`, thoughts?)

spicyj commented:
I don't really want more versions. One per language already feels like a lot to me.

Zopieux commented:
@zpao 
>  Does pip have anything to describe dependencies, like package.json for npm and Gemfile for ruby?

Use a [`requirements.txt`](https://pip.pypa.io/en/latest/user_guide.html#requirements-files) file with the single line `flask`. I would also suggest to recommand using a virtualenv.

dbunker commented:
Adding a requirements.txt might be overkill for something this simple, `pip install flask` and `gem install sinatra` could be enough. But would be happy to add it if we want it to match up to the node server's use of npm and package.json.

spicyj commented:
Yes, let's have a requirements.txt. Ideally with a version number too.

Zopieux commented:
I don't see the point of freezing the flask version given the simplicity of the example code. Remember you prevent users from getting useful updates when freezing.

spicyj commented:
You also prevent your code from breaking if Flask makes breaking API changes. I agree it doesn't make a big difference for a small script.
